server/server.go

```diff
@@ -32,55 +62,170 @@ func StartHttpServer(opts *options.Options) error {
-			} else {
-				logger.Infof("TCP connection to port %d successful", port)
+				logger.Warnf(err.Error())
+				atomic.AddUint64(&failedChecks, 1)
 			}
-
-			waitGroup.Done()
-		}(port)
+		}(check)
 	}
 
 	waitGroup.Wait()
 
-	if allPortsValid {
-		logger.Infof("All health checks passed. Returning HTTP 200 response.\n")
-		return &httpResponse{ StatusCode: http.StatusOK, Body: "OK" }
-	} else {
+	failedChecksFinal := atomic.LoadUint64(&failedChecks)
+	if failedChecksFinal > 0 {
 		logger.Infof("At least one health check failed. Returning HTTP 504 response.\n")
-		return &httpResponse{ StatusCode: http.StatusGatewayTimeout, Body: "At least one health check failed" }
+		return &httpResponse{StatusCode: http.StatusGatewayTimeout, Body: "At least one health check failed"}
+	} else {
+		logger.Infof("All health checks passed. Returning HTTP 200 response.\n")
+		return &httpResponse{StatusCode: http.StatusOK, Body: "OK"}
+	}
+}
+
+func (c TcpCheck) ValidateCheck () error {
+	if c.Name == "" {
+		return &InvalidCheck{name: "tcp", key: "name"}
+	}
+	if c.Host == "" {
+		return &InvalidCheck{name: c.Name, key: "host"}
 	}
+	if c.Port == 0 {
+		return &InvalidCheck{name: c.Name, key: "port"}
+	}
+	return nil
 }
 
-// Attempt to open a TCP connection to the given port
-func attemptTcpConnection(port int, opts *options.Options) error {
+func (c TcpCheck) DoCheck (opts *options.Options) error {
 	logger := opts.Logger
-	logger.Infof("Attempting to connect to port %d via TCP...", port)
+	logger.Infof("Checking %s at %s:%d via TCP...", c.Name, c.Host, c.Port)
 
-	defaultTimeout := time.Second * 5
+	timeout := time.Second * DEFAULT_CHECK_TIMEOUT
+	if c.Timeout != 0 {
+		// override default with user defined timeout
+		timeout = time.Second * time.Duration(c.Timeout)
+	}
 
-	conn, err := net.DialTimeout("tcp", fmt.Sprintf("0.0.0.0:%d", port), defaultTimeout)
+	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%d", c.Host, c.Port), timeout)
 	if err != nil {
-		return err
+		return &CheckFail{name: c.Name, reason: err.Error()}
 	}
 
 	defer conn.Close()
+	logger.Infof("Check SUCCESS: %s", c.Name)
+	return nil
+}
 
+func (c HttpCheck) ValidateCheck () error {
+	if c.Name == "" {
+		return &InvalidCheck{name: "http", key: "name"}
+	}
+	if c.Host == "" {
+		return &InvalidCheck{name: c.Name, key: "host"}
+	}
+	if c.Port == 0 {
+		return &InvalidCheck{name: c.Name, key: "port"}
+	}
+	return nil
+}
+
+func (c HttpCheck) DoCheck (opts *options.Options) error {
+	logger := opts.Logger
+	logger.Infof("Checking %s at %s:%d via HTTP...", c.Name, c.Host, c.Port)
+
+	timeout := time.Second * DEFAULT_CHECK_TIMEOUT
+	if c.Timeout != 0 {
+		// override default with user defined timeout
+		timeout = time.Second * time.Duration(c.Timeout)
+	}
+
+	client := http.Client{
+		Timeout: timeout,
+	}
+	resp, err := client.Get(fmt.Sprintf("http://%s:%d", c.Host, c.Port))
+	if err != nil {
+		return &CheckFail{name: c.Name, reason: err.Error()}
+	}
+
+	switch {
```

Comment:
Seems like instead of `switch`, it should just be a series of `if` statements: if there are success status codes specified, check those; if there is a body regex, check that. In fact, if success status codes aren't specified, it seems intuitive to check for 200 OK, so there's always something to check.